tinygrad/runtime/ops_webgpu.py

```diff
@@ -2,62 +2,280 @@
-    bindings += [{"binding": i+1, "resource": {"buffer": create_uniform(self.dev, x) if i >= len(bufs) else x, "offset": 0,
-                                            "size": 4 if i >= len(bufs) else x.size}} for i,x in enumerate(bufs+vals)]  # noqa: E501
-    bind_group_layout = self.dev.create_bind_group_layout(entries=binding_layouts)
-    pipeline_layout = self.dev.create_pipeline_layout(bind_group_layouts=[bind_group_layout])
-    bind_group = self.dev.create_bind_group(layout=bind_group_layout, entries=bindings)
-    compute_pipeline = self.dev.create_compute_pipeline(layout=pipeline_layout,compute={"module": self.prg, "entry_point": self.name},)
-    command_encoder = self.dev.create_command_encoder()
+    tmp_bufs = [*bufs]
+    buf_patch = False
+
+    # WebGPU does not allow using the same buffer for input and output
+    for i in range(1, len(bufs)):
+      if bufs[i] == bufs[0]:
+        tmp_bufs[0] = webgpu.wgpuDeviceCreateBuffer(self.dev,
+          webgpu.WGPUBufferDescriptor(size=webgpu.wgpuBufferGetSize(bufs[0]), usage=webgpu.wgpuBufferGetUsage(bufs[0])))
+        buf_patch = True
+
+    binding_layouts = [webgpu.WGPUBindGroupLayoutEntry(binding=0, visibility= webgpu.WGPUShaderStage_Compute,
+      buffer=webgpu.WGPUBufferBindingLayout(type=webgpu.WGPUBufferBindingType_Uniform))]
+    binding_layouts += [webgpu.WGPUBindGroupLayoutEntry(binding=i+1, visibility=webgpu.WGPUShaderStage_Compute,
+      buffer=webgpu.WGPUBufferBindingLayout(type=webgpu.WGPUBufferBindingType_Uniform if i >= len(tmp_bufs)
+      else webgpu.WGPUBufferBindingType_Storage)) for i in range(len(tmp_bufs)+len(vals))]
+    bindings = [webgpu.WGPUBindGroupEntry(binding=0, buffer=create_uniform(self.dev, float('inf')), offset=0, size=4)]
+    bindings += [webgpu.WGPUBindGroupEntry(binding=i+1, buffer=create_uniform(self.dev, x) if i >= len(tmp_bufs) else x, offset=0,
+      size=4 if i >= len(tmp_bufs) else webgpu.wgpuBufferGetSize(x)) for i,x in enumerate(tuple(tmp_bufs)+vals)]
+
+    # Creating bind group layout
+    bl_arr_type = webgpu.WGPUBindGroupLayoutEntry * len(binding_layouts)
+    webgpu.wgpuDevicePushErrorScope(self.dev, webgpu.WGPUErrorFilter_Validation)
+    bind_group_layouts = [webgpu.wgpuDeviceCreateBindGroupLayout(self.dev, webgpu.WGPUBindGroupLayoutDescriptor(
+      entryCount=len(binding_layouts), entries=ctypes.cast(bl_arr_type(*binding_layouts), ctypes.POINTER(webgpu.WGPUBindGroupLayoutEntry))))]
+    bind_group_layout_error = pop_error(self.dev)
+
+    if bind_group_layout_error: raise RuntimeError(f"Error creating bind group layout: {bind_group_layout_error}")
+
+    # Creating pipeline layout
+    pipeline_layout_desc = webgpu.WGPUPipelineLayoutDescriptor()
+    pipeline_layout_desc.bindGroupLayoutCount = len(bind_group_layouts)
+    bind_group_array_type = webgpu.WGPUBindGroupLayout * len(bind_group_layouts)
+    pipeline_layout_desc.bindGroupLayouts = bind_group_array_type(*bind_group_layouts)
```

Comment:
A nit, but I think it's more readable like this

`pipeline_layout_desc.bindGroupLayouts = (webgpu.WGPUBindGroupLayout * len(bind_group_layouts))(*bind_group_layouts)`